website: Add kconfig build system document

rogerchang23424 opened the pull request:
This is the document for Kconfig build system in the future gem5

rogerchang23424 committed:
website: Add kconfig build system document

Change-Id: I9c99aa2776676810c949d5c8bd25ba47b96417d8
Co-authored-by: Gabe Black <gabeblack@google.com>

rogerchang23424 commented:
cc:  @wmin0 @EarlOu @gabemblack

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,198 @@
+
+For advanced users who need to build gem5 with multiple ISAs or customized the
+build options, such as ruby memory protocol. Users need to familiar with
+Kconfig system.
```
```suggestion
This guide is intended for advanced users who need to build GEM5 with multiple ISAs or customize the build options, such as the Ruby memory protocol. Familiarity with the Kconfig system is required.
```

This is what an LLM suggested :). I used bing, but generally I would suggest using some LLM (e.g., bing, chatGPT, etc.) to improve the grammar of this file. I asked it "Improve the following documentation" and the copied in the raw markdown.

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,198 @@
+For advanced users who need to build gem5 with multiple ISAs or customized the
+build options, such as ruby memory protocol. Users need to familiar with
+Kconfig system.
+
```
It would be good to include a pointer the kernel kconfig docs if they cover the same information as the way we use kconfig in gem5. If the kernel implementation is different then... well that would be confusing :)

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
````diff
@@ -0,0 +1,198 @@
+scons defconfig gem5_build build_opts/ALL
+```
+
+This will create a configuration for the gem5_build build directory based on
````
Where is this configuration created? `build/ALL`? `gem5_build`? Is it in the `cwd`?

rogerchang23424 replied:
The will create the config file in `{build_dir}/gem5.build/config`. For example,  configuration for `build/ALL` will be in `build/ALL/gem5.build/config`, configuration for `gem5_build` will be in `gem5_build/gem5.build/config`

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,198 @@
+### defconfig
+
+Set up a config using values specified in a defconfig file, or if no value is
+given, use the default. The second argument specifies the defconfig file. All
```
Where would I find the default value? How is this specified?

rogerchang23424 replied:
Added in the doc description

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,198 @@
+given, use the default. The second argument specifies the defconfig file. All
+of available gem5 defconfig files are in `build_opts` directory.
+
+A defconfig file in the build_opts directory can be implicitly specified in
+the build path via `build/<defconfig file>/`
```
Could you expand on what you mean by "implicitly specified"?

rogerchang23424 replied:
Removed

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
````diff
@@ -0,0 +1,198 @@
+The user can also specify the defconfig FILE outside the gem5 repository.
+
+```bash
+scons defconfig gem5_build build_opts/RISCV
````
Is this an example of what you're describing on line 78? I don't understand what the purpose of this example is.

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,198 @@
+
+### setconfig
+
+Set values in an existing config directory as specified on the command line.
```
How would one know what variables are available to set? I don't really understand the use case here.

rogerchang23424 replied:
Please use `menuconfig` and `guiconfig` to list all of config options

powerjg replied:
This information should be in the documenation.

Apologies if my prior review wasn't clear... any "questions" I posed were because they are questions users/developers might ask that are not clear in the documentation.

When writing documentation for 100s-1000s or even 10,000s of people you have to be very clear and precise. You may also have to answer a lot of obvious (to you) questions :smile:

rogerchang23424 replied:
Added in the doc

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,198 @@
+
+![](/assets/img/kconfig/guiconfig.png)
+
+Open guiconfig in Ubuntu 22.04 with python3-tk
```
Is `python3-tk` a requirement? should this go in the `optional_requirements.txt` file? Are there any other new dependencies?

rogerchang23424 replied:
It the apt install target. I will add `python3-tk` in dockerfile.

powerjg replied:
What dockerfile? The normal use case for gem5 is not using docker.

I think this should be in the `optional-requirements.txt` file if it is not part of a default python install.

rogerchang23424 replied:
I thinks there is no ways to install `python3-tk` via pip. For Ubuntu users, they should install `python3-tk` via `apt`.
https://www.pythonguis.com/installation/install-tkinter-linux/

powerjg reviewed (commented):
This is a good start. There are a couple of things that are missing.

1. Most importantly, how do developers add new options, modify options, remove options, etc. 
2. This needs to be linked from the `how to build gem5` page. https://www.gem5.org/documentation/general_docs/building
3. You may need to modify the "getting started" page here: https://www.gem5.org/getting_started/

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,201 @@
+permalink: /documentation/general_docs/kconfig_build_system/
+---
+
+This guide is intended for advanced users who need to build GEM5 with
```
gem5 should always be lowercase

```suggestion
This guide is intended for advanced users who need to build gem5 with
```

powerjg commented on _pages/documentation/general_docs/kconfig_build_system.md:
````diff
@@ -0,0 +1,201 @@
+scons [OPTIONS] Kconfig_command TARGET
+```
+
+Supported kconfig commands include
````
Looking at the Kconfig documentation, the capitalization should always be "Kconfig" capital K lowercase config.

I know these seem like nitpicks, but if we don't get them right the first time, we'll never get them right :smile:

Harshil2107 commented:
Hi,
I am following the documentation to build using kconfig. 

I cloned the repo: https://github.com/rogerchang23424/gem5/tree/develop-kconfig from where the change is being merged.

I ran the command: `scons defconfig gem5_build build_opts/ALL`

It gives me the following error: `Error: No non-leaf 'build' dir found on target path. /home/harshilp/gem5/defconfig`

Do you know how I can fix this error? Are there any dependencies I am missing?

Harshil2107 commented on _pages/documentation/general_docs/kconfig_build_system.md:
````diff
@@ -0,0 +1,203 @@
+to before.
+
+```bash
+scons gem5_build/gem5.opt
````
Could we include a reference to the `-j [number of threads]` option in our documentation? While I was adhering to the documentation, I initially initiated the build process using a single thread. It would be beneficial to explicitly state that users have the option to specify the number of threads for the build.

Harshil2107 commented on _pages/documentation/general_docs/kconfig_build_system.md:
````diff
@@ -0,0 +1,203 @@
+to before.
+
+```bash
+scons gem5_build/gem5.opt
````
Could we add a note regarding the updated path to the `gem5.opt` file? While I was following the documentation, post-building, I attempted to execute a script using `./build/ALL/gem5.opt`, which resulted in an error. Subsequently, the operation succeeded upon using `./gem5_build/gem5.opt`.

rogerchang23424 replied:
Added in the following line of this

ivanaamit commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,206 @@
+- olddefconfig
+- savedefconfig
+
+The most common use of options are `defconfig`, `setconfig` and `menuconfig`.
```
suggested grammar change:
use --> uses

ivanaamit commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,206 @@
+`build_opts/ALL`. The configuration is stored in the
+`gem5_build/gem5.build/config`.
+
+To build a target in a configured build directory, run a scons command similar
```
suggested grammar change: 
To build a target in a configured build directory, run a scons command similar to the one used before.

ivanaamit commented on _pages/documentation/general_docs/kconfig_build_system.md:
````diff
@@ -0,0 +1,206 @@
+```
+
+Note: In order to have backward compatibility for old build scheme. The users
+need to avoid use "build" directory for Kconfig build.
````
suggested grammar change: 
Note: In order to maintain backward compatibility with the old build scheme, users need to avoid using the "build" directory for Kconfig builds.

ivanaamit commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,206 @@
+Note: In order to have backward compatibility for old build scheme. The users
+need to avoid use "build" directory for Kconfig build.
+
+To build gem5 Kconfig with customized Kconfig options, the additional step
```
suggested grammar change: 
the additional step --> an additional step

ivanaamit commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,206 @@
+
+Set up a config using values specified in a defconfig file, or if no value is
+given, use the default. The second argument specifies the defconfig file. All
+of default gem5 defconfig files are in `build_opts` directory. The user can
```
suggested grammar change: 
All default gem5 defconfig files are located in the `build_opts` directory. Users can also use their own defconfig files.

ivanaamit commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,206 @@
+
+### menuconfig
+
+Opens the menuconfig editor which will let you view and edit config values,
```
suggested grammar change: 
Opens the menuconfig editor, which allows you to view and edit config values and view help text.

ivanaamit commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,206 @@
+
+![](/assets/img/kconfig/menuconfig_details.png)
+
+If the `gem5_build` directory is not exists. SCons will set up a build
```
suggested grammar change:
If the `gem5_build` directory does not exist, SCons will set up a build directory at the path `gem5_build` with default options and then invoke menuconfig so you can set up its configuration.

ivanaamit commented on _pages/documentation/general_docs/kconfig_build_system.md:
```diff
@@ -0,0 +1,206 @@
+config how you want it with the options you cared about, and then use
+savedefconfig to save a minimal config file. These files would be suitable to
+use in the build_opts directory. The second argument specifies the filename
+for the new defconfig file.
```
suggested grammar change: 
This helper utility allows you to save the defconfig that corresponds to a given configuration. For instance, you can use menuconfig to set up a configuration with the options you care about, and then use savedefconfig to create a minimal configuration file. These files are suitable for use in the `build_opts` directory. The second argument specifies the filename for the new defconfig file.

rogerchang23424 commented:
> This needs to be linked from the how to build gem5 page. https://www.gem5.org/documentation/general_docs/building

See: https://github.com/gem5/website/pull/86

> You may need to modify the "getting started" page here: https://www.gem5.org/getting_started/

I think the page is for typical user and is not needed to update

rogerchang23424 committed:
website: Update build commands to kconfig

1. Update part1_1_building document
2. Update documentation of MSIBuilding step for kconfig
3. Update Garnet Synthetic Traffic document

The part1.1 and part3.7 is moved from gerrit:
https://gem5-review.googlesource.com/c/public/gem5-website/+/58749

powerjg reviewed (commented):
This looks OK to me. 

There's a minor problem... If we update the website with this information, it will apply to develop, but it will be *wrong* with respect to the stable release. We currently don't have a good way to explain things to users when there's a difference between develop and stable.

rogerchang23424 commented:
Thanks for the reply. Perhaps the webpage should add version annotation like `Work for gem5 >= 23.1` for the kconfig webpage?

If the Kconfig gem5/gem5#69 is merge. The [building MSI](https://www.gem5.org/documentation/learning_gem5/part3/MSIbuilding/) step of learning_gem5 part3 may need to be changed because the original approach of `Compiling a protocol with SCons` will not work anymore. The page need to be updated.

powerjg commented:
> If the Kconfig https://github.com/gem5/gem5/pull/69 is merge. The [building MSI](https://www.gem5.org/documentation/learning_gem5/part3/MSIbuilding/) step of learning_gem5 part3 may need to be changed because the original approach of Compiling a protocol with SCons will not work anymore. The page need to be updated.

Thanks for pointing this out. Can you please update it?

> Thanks for the reply. Perhaps the webpage should add version annotation like Work for gem5 >= 23.1 for the kconfig webpage?

Yes, that would help. Can you make this change?

BobbyRBruce reviewed (dismissed):
I may jump in here at some point and add some patches to rephrase some things here and there, but the content is all there.

While I approve, we need to wait until the v23.1 release to merge this into stable. So, I suppose, on your end @rogerchang23424: Just consider this "Done" and I'll handle the rest.

rogerchang23424 commented:
bump

BobbyRBruce committed:
website: Improved some grammer and phrasing for kconfigs
